addons/purchase/controllers/portal.py

```diff
@@ -111,6 +110,34 @@ def portal_my_purchase_order(self, order_id=None, access_token=None, **kw):
 
         report_type = kw.get('report_type')
         if report_type in ('html', 'pdf', 'text'):
-            return self._show_report(model=order_sudo, report_type=report_type, report_ref='purchase.action_report_purchase_order',download=kw.get('download'))
+            return self._show_report(model=order_sudo, report_type=report_type, report_ref='purchase.action_report_purchase_order', download=kw.get('download'))
+
+        confirm_type = kw.get('confirm')
+        if confirm_type == 'reminder':
+            order_sudo.confirm_reminder_mail()
+        if confirm_type == 'purchase_order':
+            order_sudo.confirm_purchase_order_mail()
+
         values = self._purchase_order_get_page_view_values(order_sudo, access_token, **kw)
-        return request.render("purchase.portal_my_purchase_order", values)
+        update_date = kw.get('update')
+        if update_date == 'True':
+            return request.render("purchase.portal_my_purchase_order_update_date", values)
+        else:
+            return request.render("purchase.portal_my_purchase_order", values)
+
+    @http.route(['/my/purchase/<int:order_id>/update'], type='http', methods=['POST'], auth="public", website=True)
+    def portal_my_purchase_order_update_dates(self, order_id=None, access_token=None, **kw):
+        order_sudo = request.env['purchase.order'].sudo().browse(order_id).exists()
+
+        updated_dates = []
+        try:
+            for id_str, date in kw.items():
+                line = order_sudo.order_line.filtered(lambda l: l.id == int(id_str))
+                updated_dates.append((line, date))
+        except Exception:
```

Comment:
what kind of exceptions could raise here ? and don't you need to check the access token?

Reply:
not sure for the exception again, but maybe id_str doesn't match any line, and that's a case you should handle